### ISSUE TITLE
Struct return transformation: function declaration

### ISSUE THREAD

**Issue by mbbarbosa** [AbsInt/CompCert#29]
The code in lines 52-59 of cparser/StructReturn.ml does not seem to correctly re-write function declarations of the form

`struct a func(void);`

The re-written output is:

`extern void f(void);`

When I guess it should be:

`extern void f(struct a *);`

**Comment by bschommer** [comment]
I had a quick look at the current master and it seems to do the right thing.

For example:

```struct a foo(void)```

is translate either to

```unsigned int foo(void)```

if a fits in an integer of

```void foo(strucht* a res)```

**Comment by mbbarbosa** [comment]
To clarify, I add my input program:

```
struct foo {
   int afield[10];
};

struct foo afunction();

int anotherf() {
   struct foo var = afunction();
   return var.afield[0];
} 
```

And an excerpt of the .compcert.c file.

```
struct foo;
struct foo {
  int afield[10];
};

extern void afunction(void);

int anotherf(void)
{
  struct foo var;
  afunction(&var);
  return *(var.afield + 0);
} 
```

**Comment by bschommer** [comment]
I think the problem is more a problem of the printing. The declaration ot the function afunction
```struct foo afunction() ```
is unprototyped and in cparser/StructReturn.ml it is again translated to an unprototyped function. However in 
If you declare the function as:
```struct foo afunction();```
you get the expected result.

**Comment by xavierleroy** [comment]
One way to solve this mystery is to add the "-dparse" option and look at the .parsed.c output.  It contains the source C code after transformations such as StructReturn but before conversion to the CompCert C subset.  On the example above, here are the contents of .parsed.c:

    struct foo {
       int afield[10];
    };

    struct foo afunction();

    int anotherf() {
       struct foo var = afunction();
       return var.afield[0];
    } 

As you can see, the unprototyped declaration of "afunction" remains unprototypedafter StructReturn conversion.

In the CompCert C subset of C (what you get in .compcert.c files with the "-dc" option), all functions must have a prototype: the formal semantics are undefined otherwise.  Indeed, if you compile in strict mode (-fnone), any unprototyped function declaration is an error.  As a tolerance, in non-strict mode, unprototyped function declarations are turned into functions with zero arguments, printed as "(void)", and the CompCert compiler manages to produce correct code for calls to these functions, even though they fall outside the formal semantics.

In summary: 
- I believe StructReturn is doing the right transformation here.
- Unprototyped functions are iffy and should not be used, ever.
  (Function prototypes were introduced in 1989, for heaven's sake.)
- Berhard is right that the printing of the .compcert.c file could
  possibly be improved by marking unprototyped function declarations
  specially and printing them as "()" rather than "(void)".

**Comment by mbbarbosa** [comment]
Thanks very much for the clarification.

**Comment by xavierleroy** [comment]
Commit [master ef16a51]: better printing of unprototyped function declarations in .compcert.c files.
